### PR TITLE
refactor(ethexe): Subscribe blocks when `Observer` is created

### LINKED ISSUE #4099: ethexe: append possibility to receive info that service is started to listen blocks
### Problem to Solve

Presently we have no ability to understand that service has run blocks listening in tests. That cause the problem that we need to sleep 1 second and hope that it's enough for service to start listening.

### Possible Solution

Maybe it's normal to receive event "StartListening" thru network.

### Notes

_No response_

### PR DESCRIPTION
**Pull request by techraed**
Resolves #4099 

TODO:
- [x] use macro to avoid copy-paste in `EventsStreamProducer`

**UPD**:
Still partially resolves #4099

### CODE REVIEW

**Comment by techraed** [comment]
I'd say that tests are really fragile: removing timers results in underlying spawned tasks being not synchronized, which creates many obstacles in testing. The approach to test impl should be changed in future, so we can write them easily without thinking too much about underlying details of how `Service` and other entities are implemented.

 [6e450c4](https://github.com/gear-tech/gear/pull/4311/commits/6e450c46487c1b2a372002ad7b667cb973a38186) - this works perfectly fine, but requires some timeout for sync.

[a465d72](https://github.com/gear-tech/gear/pull/4311/commits/a465d72aa3ceae68f3f5a8705bf56829f96ca095) - this avoids sync problems when channgels try to read data, when there's no actual data sent from the `anvil`, because it works in a non-block-production mode. However, it requires increasing a number of attempts to get the receipt for the sent tx (could be the `anvil` problem).

File: ethexe/ethereum/src/lib.rs [hunk 1]
```diff
@@ -297,7 +297,9 @@ impl<T: Transport + Clone, N: Network> TryGetReceipt<T, N> for PendingTransactio
             Err(err) => err,
         };
 
-        for _ in 0..3 {
+        log::trace!("Failed to get transaction receipt for {tx_hash}. Retrying...");
+        for n in 0..25 {
```
Comment (StackOverflowExcept1on) [inline ethexe/ethereum/src/lib.rs:301]: it's ok for tests, but bad for production

**Review by StackOverflowExcept1on — APPROVED** [review]

File: ethexe/cli/src/service.rs [hunk 2]
```diff
@@ -418,30 +420,46 @@ impl Service {
...
             mut query,
             mut router_query,
             mut processor,
-            mut sequencer,
-            signer: _signer,
             mut validator,
-            metrics_service,
             rpc,
             block_time,
+            signer: _signer,
```
Comment (StackOverflowExcept1on) [inline ethexe/cli/src/service.rs:458]:
```suggestion
```

File: ethexe/cli/src/service.rs [hunk 3]
```diff
@@ -844,6 +855,62 @@ impl Service {
...
+        let Self {
+            service,
+            events_stream_producer,
+        } = self;
+
+        service
+            .run_inner(events_stream_producer)
+            .await
+            .map_err(|err| {
+                log::error!("Service finished work with error: {:?}", err);
```
Comment (StackOverflowExcept1on) [inline ethexe/cli/src/service.rs:908]:
```suggestion
                log::error!("Service finished work with error: {err:?}");
```

**Comment by techraed** [comment]
After some discussions we came to the conclusion that having blocks automatically produced is not a good solution, so dividing the PR to the 2:
- subscribing to blocks during `Observer` instantiation
- actually solving `sleep` problem

File: ethexe/cli/src/service.rs [hunk 4]
```diff
@@ -418,30 +420,46 @@ impl Service {
...
+        if let Some(metrics_service) = self.metrics_service.take() {
+            tokio::spawn(metrics_service.run(
+                observer.get_status_receiver(),
+                self.sequencer.as_mut().map(|s| s.get_status_receiver()),
+            ));
+        }
+        let events_stream_producer = observer.events_stream_producer();
+
+        self.run_inner(events_stream_producer).await.map_err(|err| {
+            log::error!("Service finished work with error: {:?}", err);
```
[the remark this answers is not in the document]
Comment (techraed) [reply ethexe/cli/src/service.rs]: not actual

File: ethexe/ethereum/src/lib.rs (hunk 1 above)
Comment (techraed) [reply ethexe/ethereum/src/lib.rs]: not actual

File: ethexe/cli/src/service.rs (hunk 3 above)
Comment (techraed) [reply ethexe/cli/src/service.rs]: done

**Review by StackOverflowExcept1on — APPROVED** [review]

**Review by grishasobol — APPROVED** [review]

**Pull request merged**